### ISSUE TITLE
"Discard all changes" context menu on Changes header removed

### ISSUE THREAD

**Issue by varkor** [desktop/desktop#7696]
Right-clicking on the "4 changed files" header in the changed files list no longer produces a context menu. The context menus on the individual files themselves still works.

I'm on GitHub Desktop version 2.0.2 on macOS 10.12.6.

**Comment by shiftkey** [comment]
@varkor thanks for the feedback!

This was a change we decided to make in #7394 to address some UX concerns about "Discard All Changes". We had some feedback about this after #7415 was merged which I'll pull into this issue and figure out how we want to proceed.

**Comment by shiftkey** [comment]
From @fatcerberus in https://github.com/desktop/desktop/pull/7415#issuecomment-499160177

> I don't even know how I first discovered this option - I think I just subconsciously knew it would be there, which doesn't exactly say "bad UX" to me.
> 
> I'm sad this command is now gone and moved to the menu bar which isn't anywhere near as ergonomic - this was a big part of my workflow (do some experimental change -> experiment fails -> discard-all and try again). I didn't find the command particularly unsafe as there's a huge "Are you really sure?" warning that pops up which AFAICT can't be disabled, and right-clicking "X changed files" to discard everything is a natural extension of right-clicking individual files to discard them.
> 
> A keyboard shortcut would help, but personally I'm kind of mouse-bound (which is why I want to use a GUI client in the first place); right-clicking things to discover new stuff comes more naturally than hunting through menus, so I tend to miss features if that's the only way to discover them.

From @nycdotnet in https://github.com/desktop/desktop/pull/7415#issuecomment-499206333

> The old UX felt very tactile/interactive.

From @NickCraver in https://github.com/desktop/desktop/pull/7415#issuecomment-499209904

> I used this many times a day. When right clicking on an item like a file, I expect to have options for that thing. When I see a count up top, I naturally right clicked thinking I'd have options for n thing (and I did!). I was pretty surprised this was removed. Does also having the right click menu there cause some issue? Selecting all the files to do an action or hunting in another place is not intuitive to me. It also puts a lot of distance between a global work-loss action and the count of things I'm about to execute it on. That context should be readily apparent. While it did say "Discard all changes" before, it's far away from context now, and I have to go mentally check if I want to really execute that dangerous action. When it was immediately beside the things it affects? Yeah sure, I know where I'm at and what's about to happen. I know that's minor, but I wanted to convey why that feels like a regression in and of itself.

**Comment by varkor** [comment]
For the record, I agree with the quotes you've posted: though it's fine to have a main menu option too, the context menu was an intuitive and efficient place to find this option and it's a little frustrating to have to learn a new UX pattern when the old one worked well.

**Comment by AevVRC** [comment]
Was a big fan of that context menu and used it often, would love to see it return in some form.

**Comment by JamesToohey** [comment]
First thing I came to do was report a bug (assuming it was mistakenly removed). I think it was an intuitive place for this control

**Comment by evromalarkey** [comment]
Yeah, this change really seems like a step back

**Comment by Ravi5ingh** [comment]
Just leaving my desire here. I would like to have this feature. I agree with @varkor and others. Taking this feature out is a step back.

**Comment by photonstorm** [comment]
I came here to report this as a bug, but see that it was done on purpose. I honestly think calling it a UX issue was incorrect, it made complete sense for it to be there, next to the 'all files' checkbox, and was part of my daily workflow :(

**Comment by DanTup** [comment]
I also really miss this. All day I've been trying to use it (I didn't realise it had moved at the time). I think having it on the context menu of text that says "15 changed files" is a completely intuitive place to put it. Please bring it back :-)

**Comment by billygriffin** [comment]
Thanks for all the feedback folks, this is great! Sorry for making your workflows more difficult in the near term. We're discussing some options internally and expect to iterate on this - just wanted to make sure y'all knew you weren't shouting into a void, and appreciate everyone's thoughts.

**Comment by the-avid-engineer** [comment]
Might be an unpopular opinion, but I have accidentally clicked on **Discard all changes** a few times, and because of that, I like that it's not a part of the Changes context menu any more. (Fortunately discarded changes go to the trashcan instead of binary hell, but still..)

Sure, you get a confirmation modal, but you also get a confirmation modal when you discard a subset of changes. And who actually reads those after the first few times? Not me. I instinctively go straight for the confirm button.

**Comment by fatcerberus** [comment]
@the-avid-engineer There’s a reason I disable the confirmation for discarding single files :)

**Comment by NickCraver** [comment]
@the-avid-engineer I'd love to understand your use case. It's still in the context menu per-item, and in the header case it was the *only* content menu item...so I have to ask: what did you think you were clicking on? I'm at a loss for what this would be confused with, even with prompt numbness.

**Comment by the-avid-engineer** [comment]
@NickCraver I seem to recall that if you selected a subset of changes, the context menu would show both a "Discard all changes" and a "Discard {x} selected changes" option, and that I accidentally clicked on the former a few times. Is there a way to revert to 1.x to double check?

**Comment by SimonCropp** [comment]
please can we have that option back. it was sensible to have that co-located with the other "changed file options". it is not intuitive to have to go to a top menu menu for such a common action

**Comment by darrenkopp** [comment]
Use this daily. There’s a few git actions I prefer using Desktop for, and discarding individual files / all files is the one I use the most. I can understand the UX concerns, but this doesn’t have to be the SOLE place this behavior lives. 

Plaintext UI elements are fine for right click context menu as well since that’s pretty much where all right click menus exist. 

Regardless of what you decide, I’m grateful for everything y’all do!

**Comment by SimonCropp** [comment]
> Regardless of what you decide, I’m grateful for everything y’all do!

thanks to @darrenkopp for this. looking back on my response, it might come over as passive aggressive. Wasnt my intent and GH desktop is awesome

**Comment by hazendaz** [comment]
This release is extremely buggy.  So far 3 patches within days.  I don't think this was ready and it's messed up on rebases with missing ability to even discard a single file.  It was extremely intuitive before.  I've never had to read a manual on this ever.  Now I feel like I need one.  The ability to quickly discard files is just about as close to the heart of the reason to use this as it gets.  Please restore the feature.  If there is such a desire by some to lose this feature at least allow it to be configurable for us power users.

**Comment by DanTup** [comment]
I think it's a good idea not to have on the context menu for single items/selections, it's specifically in the header that I think it makes sense.

**Comment by NickCraver** [comment]
Totally agree - I’d only want to see it return on the header, having the “all” option on individual files didn’t make much sense and I see how confusion could exist there.

**Comment by shiftkey** [comment]
> it's messed up on rebases

@hazendaz would you be able to open a separate issue about this to share this experience with me? I'd love to understand what happened here...

**Comment by hazendaz** [comment]
@shiftkey If one hasn't been opened already, I'll get some screen shots and share.  The issue is on merge conflicted files (could be others).  I've experienced it now on multiple machines.  The right click on file simply no longer has discard file in that case.

**Comment by Ravi5ingh** [comment]
@shiftkey will there be a decision take on this at some point? ie. whether or not to bring it back?

**Comment by shiftkey** [comment]
@hazendaz please do open an issue. We can confirm whether its a new case to investigate, or related to some previous behaviour.

@Ravi5ingh we'll be talking as a team this week about this and other feedback from last week's release. Stay tuned for updates.

**Comment by JosephTLyons** [comment]
There was nothing wrong with having this discard functionality in a right-click menu before.  Not sure why this was removed, taking it away after having it here for this long seems odd.  Please bring it back.

**Comment by billygriffin** [comment]
Hi folks, we removed the context menu from the header because we thought it wasn't very discoverable and thought a menu item and associated keyboard shortcut would be more discoverable. For those who used it regularly and knew it was there, it was obviously very useful (as evidenced by the response on this thread 😄). Appreciate y'all weighing in, and as I mentioned previously we're discussing this internally about how to proceed. We hear you and we're planning to iterate on this, likely in the next release.

**Comment by shiftkey** [comment]
Thanks for the feedback everyone!

We decided to just restore this in #7766 and this will go out in the `2.0.4` update scheduled for today.

**Comment by billygriffin** [comment]
Hi folks, the release is out (2.0.4) with this added back. This was on me, sorry for the disruption to your workflow. Thanks again for the feedback and for your patience.

**Comment by Ravi5ingh** [comment]
Awesome. Thank you!

**Comment by fatcerberus** [comment]
I was highly amused that this was listed as a bugfix in update notification. :wink:

Thanks for bringing it back.

**Comment by JamesToohey** [comment]
Champions

**Comment by xv** [comment]
Thanks for listening and bringing it back. 🎉

**Comment by photonstorm** [comment]
Brilliant, thank you!
